### PR TITLE
feat: Optimize Column Selection

### LINKED ISSUE #850: Optimize Column Selection in InfluxRPCPlanner::tag_columns
__Use case:__

There is a function in the gPC query frontend that restricts the returned column type to tags [`InfluxRPC::restrict_to_tags`](https://github.com/influxdata/influxdb_iox/blob/main/query/src/frontend/influxrpc.rs#L363-L373)

This could be further improved by "pushing down" the restriction to tags  to the underlying chunk implementations (ReadBuffer and MutableBuffer) so they don't return more columns than necessary. @e-dard added pushdown to the ReadBuffer here: https://github.com/influxdata/influxdb_iox/pull/849

The call chain currently looks something like

```
         ┌────────────────┐            
         │   InfluxRPC    │            
         │    frontend    │            
         └────────────────┘            
                  │                    
                  ▼                    
     ┌─────────────────────────┐       
     │    `PartitionChunk`     │       
     │  and `Database` traits  │       
     └─────────────────────────┘       
                  │                    
                  ▼                    
       ┌────────────────────┐          
       │ server::Db::Chunk  │          
       └────────────────────┘          
                  │                    
     ┌────────────┼─────────────┐      
     ▼            ▼             ▼      
┌─────────┐  ┌─────────┐  ┌───────────┐
│  Read   │  │ Mutable │  │  Parquet  │
│ Buffer  │  │ Buffer  │  │(eventually│
└─────────┘  └─────────┘  └───────────┘
```

The [`InfluxRPC` frontend](https://github.com/influxdata/influxdb_iox/blob/main/query/src/frontend/influxrpc.rs#L206-L224) knows about the Influx data model (vai a call to `table_schema`), so I was imaging it creating the `Selection` and passing that down to the chunks:

```
                   ┌────────────────┐     
 Selection<'_>     │   InfluxRPC    │     
  created here     │    frontend    │     
and passed down    └────────────────┘     
                            │             
       │                    │             
       │                    ▼             
       │       ┌─────────────────────────┐
       │       │    `PartitionChunk`     │
       │       │  and `Database` traits  │
       │       └─────────────────────────┘
       │                    │             
       │                    │             
       │                    ▼             
       │         ┌────────────────────┐   
       │         │ server::Db::Chunk  │   
       │         └────────────────────┘   
       │                    │             
       │                    │             
       │                    ▼             
       ▼               ┌─────────┐        
                       │  Read   │        
                       │ Buffer  │        
                       └─────────┘        
```
(see previous discussion on https://github.com/influxdata/influxdb_iox/pull/833#discussion_r579202908 for more background, I tried to summarize the results here)

__Proposal:__
Implement "selection pushdown" for RPCPlanner::tag_clumns

__Current behaviour:__
All columns that pass a predicate are returned and then are filtered down to tag columns only in the `InfluxRPCPlanner`

__Desired behaviour:__
The `InfluxRPCPlanner` passes down a `Selection` restriction to the underlying chunks (as above).

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #850

Describe your proposed changes here.This is the implementation for my #850 in which the **column_names** functions of Mutable Buffer and Read Buffer only return columns specified in the input column selection. This PR includes:
1. Add a 'selection' parameter into column_names function of PartitionChunk trait implemented by both Mutable Buffer and Read Buffer.
2. Then implement that selection inside Mutable Buffer's column_names (the Read Buffer's had done by Edd).
3. Change the frontend's tag_keys function to set that Selection to contain only Tag columns and push down to Mutable and Read Buffer using the column_names function implemented in the 2 steps above.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: query/src/lib.rs [hunk 1]
```diff
@@ -137,6 +137,7 @@ pub trait PartitionChunk: Debug + Send + Sync {
         &self,
         table_name: &str,
         predicate: &Predicate,
+        columns: Selection<'_>,
```
Comment (NGA-TRAN) [inline query/src/lib.rs:140]: Step 1: add parameter into the PartitionChunk trait's function

File: mutable_buffer/src/chunk.rs [hunk 2]
```diff
@@ -305,6 +305,7 @@ impl Chunk {
         &self,
         table_name: &str,
         chunk_predicate: &ChunkPredicate,
+        selection: Selection<'_>,
```
Comment (NGA-TRAN) [inline mutable_buffer/src/chunk.rs:308]: Step 2: Implement the Selection inside the Mutable Buffer

File: query/src/frontend/influxrpc.rs [hunk 3]
```diff
@@ -198,25 +198,34 @@ impl InfluxRPCPlanner {
                     chunk_id = chunk.id(),
                     "finding columns in table"
                 );
-                // try and get the list of columns directly from metadata
+
```
Comment (NGA-TRAN) [inline query/src/frontend/influxrpc.rs:201]: Step 3: Have the frontend send only Tag columns to (the PartitionChunk trait) of Mutable & Read Buffer

File: query/src/test.rs [hunk 4]
```diff
@@ -496,6 +496,41 @@ impl TestChunk {
             .push(Arc::new(batch));
         self
     }
+
```
Comment (NGA-TRAN) [inline query/src/test.rs:499]: Extra Selection filer tests

**Comment by e-dard** [comment]
@NGA-TRAN looks like there are some formatting errors. The quickest way to fix this is probably to run `cargo fmt --all` and push the result up.

File: query/src/frontend/influxrpc.rs [hunk 5]
```diff
@@ -228,25 +228,29 @@ impl InfluxRPCPlanner {
...
+                // get only tag columns from metadata
+                let schema = chunk
+                    .table_schema(&table_name, Selection::All)
+                    .await
+                    .expect("to be able to get table schema");
+                let column_names: Vec<&str> = schema
+                    .tags_iter()
```
Comment (NGA-TRAN) [inline query/src/frontend/influxrpc.rs:238]: @alamb tags_iter integrated

**Comment by NGA-TRAN** [comment]
> @NGA-TRAN looks like there are some formatting errors. The quickest way to fix this is probably to run `cargo fmt --all` and push the result up.

Yes, I got it fixed. Thanks Edd

File: mutable_buffer/src/chunk.rs [hunk 6]
```diff
@@ -360,6 +361,16 @@ impl Chunk {
             }
         }
 
+        // Only return subset of these selection_cols if not all_cols
+        let mut all_cols = true;
+        let selection_cols = match selection {
```
Comment (alamb) [inline mutable_buffer/src/chunk.rs:366]: One possible follow on improvement might be to add a `Selection::contains(col_name: &str)` so the logic could be reused, but given this is the only use so far I don't think it is a big deal

File: query/src/frontend/influxrpc.rs [hunk 7]
```diff
@@ -533,18 +537,6 @@ impl InfluxRPCPlanner {
         Ok(table_names)
     }
 
-    /// removes any columns from Names that are not "Tag"s in the Influx Data
```
Comment (alamb) [inline query/src/frontend/influxrpc.rs:536]: Love it!

**Review by alamb — APPROVED** [review]
Looks good to me @NGA-TRAN  - nice work! 

I took off the `automerge` label simply so you can decide when you want to merge. When the automerge label is present, I feel like when I click "Approve" I am effectively also merging the PR, which I am not sure is what is intended. But maybe it was :) I am still getting used to the kodiak bot

File: mutable_buffer/src/chunk.rs (hunk 6 above)
Comment (NGA-TRAN) [reply mutable_buffer/src/chunk.rs:366]: Sounds good. If I have to do his again, I refactor it like this

**Comment by alamb** [comment]
🎉

**Pull request merged**